### PR TITLE
Add scripts to publish Java tracer layer. Also add code to set the correct tags in traces

### PR DESCRIPTION
**Pull request by agocs**
### What does this PR do?

This PR adds a script that publishes [dd-trace-java](https://github.com/DataDog/dd-trace-java) as a lambda layer. That will enable customers to add dd-trace-java to their lambdas as a java agent. Additionally, it also sets useful values inside the tracer to make sure the traces get back to Datadog.

### Testing Guidelines

All manual

### Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Checklist

- [ ] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog

### CODE REVIEW

File: scripts/list_layers.sh [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019 Datadog, Inc.
+
+# Lists most recent layers ARNs across regions to STDOUT
+# Optionals args: [layer-name] [region]
+
+set -e
+
+LAYER_NAMES=("Datadog-Python27" "Datadog-Python36" "Datadog-Python37" "Datadog-Python38")
```
Comment (nhinsch) [inline scripts/list_layers.sh:13]: Are these the correct layer names?

**Comment by nhinsch** [comment]
1. I would definitely add a step to sign the layer before publishing. You can copy most of the code from one of the other repos. I can help you out if you have any questions. Then I would update the documentation/README with instructions for how to use this layer with code signing (it'll be the same as the other layers).

2. I would create separate scripts intended for publishing the layer to production and to sandbox that invoke the general publish and signing scripts with specific parameters, like we do for the other layers:
- For sandbox, we should only publish to `sa-east-1`. For production, we should publish to all available regions.
- The signing script also needs to know which account we want to use to sign the layer.

3. Finally, can we also bring over the "add new region" script that we have in the other layers as well?

File: scripts/list_layers.sh (hunk 1 above)
Comment (agocs) [reply scripts/list_layers.sh]: 👍 deleted the unused line

**Comment by agocs** [comment]
Added code signing and test-layer-publishing. I'm going to defer on add new regions for now, in the interest of wanting to release this version so that @Hesperide can start testing again

File: scripts/list_layers.sh [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		LAYERS_MISSING_REGIONS+=( $region )
+	    fi
+	else
+	    echo $last_layer_arn
+	fi
+done
+
+if [ ${#LAYERS_MISSING_REGIONS[@]} -gt 0 ]; then
+    echo "WARNING: Following regions missing layers: ${LAYERS_MISSING_REGIONS[@]}"
+    echo "Please run ./add_new_region.sh <new_region> to add layers to the missing regions"
```
Comment (nhinsch) [inline scripts/list_layers.sh:49]: You mentioned that you weren't going to add the `add_new_region` script yet, so I would take this out for now

File: scripts/list_layers.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
+# Unless explicitly stated otherwise all files in this repository are licensed
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https://www.datadoghq.com/).
+# Copyright 2019 Datadog, Inc.
+
+# Lists most recent layers ARNs across regions to STDOUT
+# Optionals args: [layer-name] [region]
```
Comment (nhinsch) [inline scripts/list_layers.sh:9]: Optionals -> optional

Also, is layer-name really an arg? It looks like it's hardcoded as `dd-trace-java`

**Review by nhinsch — APPROVED** [review]
lgtm, just a couple of suggestions

**Comment by nhinsch** [comment]
I would also reiterate my suggestion of bringing over the add new region script. It's probably not too hard since it's already written, and I suspect we'll need it soon enough, so we might as well get it out of the way now. But everything here looks good.

File: scripts/sign_layers.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    S3_BUCKET_NAME="dd-lambda-signing-bucket-sandbox"
+fi
+if [ "$1" = "prod" ]; then
+    REGION="us-east-1"
+    S3_BUCKET_NAME="dd-lambda-signing-bucket"
+fi
+
+if [ -z "$2" ]
+then
+  echo "Usage: ./sign_layers (staging|prod) layer_file.zip"
```
Comment (nhinsch) [inline scripts/sign_layers.sh:33]: Should be `(sandbox|prod)`

File: scripts/sign_layers.sh [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+S3_SIGNED_ZIP_URI="s3://${S3_BUCKET_NAME}/${S3_SIGNED_ZIP_KEY}"
+aws s3 cp $S3_SIGNED_ZIP_URI $LAYER_FILE
+
+# Delete the signed and unsigned ZIPs in S3
+echo "Cleaning up the S3 bucket..."
+aws s3api delete-object --bucket $S3_BUCKET_NAME --key $S3_UNSIGNED_ZIP_KEY
+aws s3api delete-object --bucket $S3_BUCKET_NAME --key $S3_SIGNED_ZIP_KEY
+
+echo
+echo "Successfully signed all layers!"
```
Comment (nhinsch) [inline scripts/sign_layers.sh:108]: all layers -> the layer (for this repo)

File: src/main/java/com/datadoghq/datadog_lambda_java/DDLambda.java [hunk 6]
```diff
@@ -1,150 +1,217 @@
...
+        String functionName = "";
+        String functionArn = "";
+        if (cxt != null) {
+            requestId = cxt.getAwsRequestId();
+            functionName = cxt.getFunctionName();
+            functionArn = santitizeFunctionArn(cxt.getInvokedFunctionArn());
+        }
+        Span span = GlobalTracer.get().activeSpan();
+        if (span != null) {
+            span.setTag("request_id", requestId);
```
Comment (nhinsch) [inline src/main/java/com/datadoghq/datadog_lambda_java/DDLambda.java:120]: I think the other libraries add `function_version` and `resource_names` tags as well.

Comment (agocs) [reply src/main/java/com/datadoghq/datadog_lambda_java/DDLambda.java:122]: 👍

**Pull request merged**